OSDOCS#9497: editing canary rollout update page (#71004)
openshift/openshift-docs: updating/updating_a_cluster/update-using-custom-machine-config-pools.adoc

To do: Remove this comment once 4.13 docs are EOL.
////

A _canary update_ is an update strategy where worker node updates are performed in discrete, sequential stages instead of updating all worker nodes at the same time.
This strategy can be useful in the following scenarios:

* You want a more controlled rollout of worker node updates to ensure that mission-critical applications stay available during the whole update, even if the update process causes your applications to fail.
* You want to update a small subset of worker nodes, evaluate cluster and workload health over a period of time, and then update the remaining nodes.
* You want to fit worker node updates, which often require a host reboot, into smaller defined maintenance windows when it is not possible to take a large maintenance window to update the entire cluster at one time.

In these scenarios, you can create multiple custom machine config pools (MCPs) to prevent certain worker nodes from updating when you update the cluster.
After the rest of the cluster is updated, you can update those worker nodes in batches at appropriate times.

[id="example_{context}"]
== Example Canary update strategy

> The following example describes a canary update strategy where you have a cluster with 100 nodes with 10% excess capacity, you have maintenance windows that must not exceed 4 hours, and you know that it takes no longer than 8 minutes to drain and reboot a worker node.

Reviewer: @skopacz1 How did you come up with 8 minutes for each node drain out as an example?

Reviewer: in https://docs.openshift.com/container-platform/4.14/updating/understanding_updates/understanding-openshift-update-duration.html we have 
```
The time to complete this process depends on several factors including the node and infrastructure configuration. This process might take 5 or more minutes to complete per node.
```

Author: This is existing content from the current page, as can be seen [here](https://docs.openshift.com/container-platform/4.14/updating/updating_a_cluster/update-using-custom-machine-config-pools.html#:~:text=For%20example%2C%20if%20you%20have%20a%20cluster%20with%20100%20nodes%20with%2010%25%20excess%20capacity%2C%20maintenance%20windows%20that%20must%20not%20exceed%204%20hours%2C%20and%20you%20know%20that%20it%20takes%20no%20longer%20than%208%20minutes%20to%20drain%20and%20reboot%20a%20worker%20node%2C).

I'm guessing that the person who added this content originally just chose an example number for the sake of demonstrating how they could plan a canary rollout strategy given certain factors, and I don't think they were suggesting users create an expectation of their nodes having the same exact drain time. 

If you'd like to replace that value and help me update the rest of the example here, I'd be happy to do that. Or I can add some note about these only being example values and how your cluster may perform differently.

Reviewer 2: > I can add some note about these only being example values and how your cluster may perform differently.

Agree with ^. The time to drain a node may vary depending on how much user workload there is.

.Defining custom machine config pools

In order to organize the worker node updates into separate stages, you could begin by defining the following MCPs:

* *workerpool-canary* with 10 nodes
* *workerpool-A* with 30 nodes
* *workerpool-B* with 30 nodes
* *workerpool-C* with 30 nodes

.Updating the canary worker pool

During your first maintenance window, you would pause the MCPs for *workerpool-A*, *workerpool-B*, and *workerpool-C*, and then initiate the cluster update.
This updates components that run on top of {product-title} and the 10 nodes that are part of the *workerpool-canary* MCP, because that pool was not paused.
The other three MCPs are not updated, because they were paused.

.Determining whether to proceed with the remaining worker pool updates

> If for some reason you determine that your cluster or workload health was negatively affected by the *workerpool-canary* update, you would then cordon and drain all nodes in that pool while still maintaining sufficient capacity until you have diagnosed the problem.

Reviewer 2: I'm trying to understand the statement. If there's a negative impact on the cluster or workload health due to the update of the workerpool-canary, we are suggesting users to cordon and drain all nodes in the workerpool-canary so that the workload will be migrated to nodes in other work pools. Right? 

@jianlinliu @rioliu-rh PTAL. Thank you.

Reviewer 3: I think the issue here is, when the update is stuck at a node (`maxUnavailable = 1`), the pool is degraded, we need to do RCA to figure out the root cause and apply a fix on it, then rest of nodes will be updated (cordon, drain, update, reboot etc) automatically by MCO

Reviewer: The documentation is suggesting that if the user find out that the new update is not working for the application after updating the first machine config pool, the user can cordon and drain the application from nodes with new version. Because the cluster has sufficient capacity the quality of service wont be impacted. Also it is assumed that update is successful but the application might not be compatible with the worker nodes with the new version because of some API incompatibility etc.

Author: Is there any doc change that should be incorporated here? Perhaps making it more explicit that we expect users to resolve the issue before resuming their update? Maybe the following:
```suggestion
If for some reason you determine that your cluster or workload health was negatively affected by the *workerpool-canary* update, you would then cordon and drain all nodes in that pool while still maintaining sufficient capacity until you have diagnosed and resolved the problem.
Once the issue is resolved, you would then uncordon the nodes to have workloads scheduled to them again.
```

Reviewer 2: The update^ looks good to me.

In these scenarios, you can create multiple custom machine config pools (MCPs) to prevent certain worker nodes from updating when you update the cluster.
After the rest of the cluster is updated, you can update those worker nodes in batches at appropriate times.

[id="example_{context}"]
== Example Canary update strategy

The following example describes a canary update strategy where you have a cluster with 100 nodes with 10% excess capacity, you have maintenance windows that must not exceed 4 hours, and you know that it takes no longer than 8 minutes to drain and reboot a worker node.

[NOTE]
====
The previous values are an example only.
The time it takes to drain a node might vary depending on factors such as workloads.
====

.Defining custom machine config pools

In order to organize the worker node updates into separate stages, you could begin by defining the following MCPs:

* *workerpool-canary* with 10 nodes
* *workerpool-A* with 30 nodes
* *workerpool-B* with 30 nodes
* *workerpool-C* with 30 nodes

> .Updating the canary worker pool

Author: I can do that. My main intention was to segment these chunks of content with some sort of formatted text, but without over-saturating the right-side nav of the assembly. So I can make these level 2 subheadings, but I'll make them [discrete] since they're not necessary to include in the navigation.

> .Updating the canary worker pool

Reviewer 4: Someone mentioned you might make them discrete, then said: 

I agree, those should be L2s. If the author is trying to make the TOC uncluttered, they could make them [discrete], but I think it's better for the user to have those in the TOC anyway.

So in a merge review you might get pushback for it. I don't have a huge preference; I think the ToC could maybe _look_ better with them. up to you.

> .Updating the canary worker pool

Author: I have a slight preference to keep them discrete so I'll see what the merge reviewer says. If they need to have the discrete attribute removed then that would be fine with me as well.

> During your first maintenance window, you would pause the MCPs for *workerpool-A*, *workerpool-B*, and *workerpool-C*, and then initiate the cluster update.

Author: I can remove "would" although I'm not sure "must" is the right replacement here. This entire section is meant to be descriptive and walk a user through an example implementation of this strategy, rather than describe what a user must do in that instance, if that makes sense. How about I remove would but don't replace it with anything, like the following:
```suggestion
During your first maintenance window, you pause the MCPs for *workerpool-A*, *workerpool-B*, and *workerpool-C*, and then initiate the cluster update.
```

> During your first maintenance window, you would pause the MCPs for *workerpool-A*, *workerpool-B*, and *workerpool-C*, and then initiate the cluster update.

Reviewer 4: Works better as present tense like this anyways.

> This updates components that run on top of {product-title} and the 10 nodes that are part of the *workerpool-canary* MCP, because that pool was not paused.

Author: I see what you mean. This sentence is a bit clunky the more I read it, because it's saying the following:
"This updates:
1.  components that run on top of {product-title}
2. the 10 nodes that are part of the *workerpool-canary* MCP (because that pool was not paused)"

Would this sound better or make it even harder to parse?
```suggestion
This updates components that run on top of {product-title} and the 10 nodes that are part of the unpaused *workerpool-canary* MCP.
```

The next sentence explains how paused vs unpaused pools affect whether or not the pool is updated, so I don't think it's necessary to explain that here.

> This updates components that run on top of {product-title} and the 10 nodes that are part of the *workerpool-canary* MCP, because that pool was not paused.

Author: I'll also remove the comma before "because" in the next sentence, since I agree that it's not helpful.

The other three MCPs are not updated, because they were paused.

.Determining whether to proceed with the remaining worker pool updates

> If for some reason you determine that your cluster or workload health was negatively affected by the *workerpool-canary* update, you would then cordon and drain all nodes in that pool while still maintaining sufficient capacity until you have diagnosed and resolved the problem.

Author: Removing would but not replacing with must (unless you've determined in the above comment that "must" is still better to include)

> When everything is working as expected, you would then evaluate the cluster and workload health before deciding to unpause, and thus update, *workerpool-A*, *workerpool-B*, and *workerpool-C* in succession during each additional maintenance window.

Author: Removing would but not replacing with must (unless you've determined in the above comment that "must" is still better to include)

This scenario has not been tested and might result in an undefined cluster state.
====

[id="update-using-custom-machine-config-pools-about-mcp_{context}"]
== About the canary rollout update process and MCPs

In {product-title}, nodes are not considered individually, instead they are grouped into machine config pools (MCP).
By default, nodes in an {product-title} cluster are grouped into two MCPs: one for the control plane nodes and one for the worker nodes.
An {product-title} update affects all MCPs concurrently.

During the update, the Machine Config Operator (MCO) drains and cordons all nodes within a MCP up to the specified `maxUnavailable` number of nodes (if specified), by default `1`. Draining and cordoning a node deschedules all pods on the node and marks the node as unschedulable.

After the node is drained, the Machine Config Daemon applies a new machine configuration, which can include updating the operating system (OS). Updating the OS requires the host to reboot.

.Using custom machine config pools

To prevent specific nodes from being updated, you can create custom MCPs.

> Because the MCO does not update nodes within paused MCPs, you would then pause the MCPs containing nodes that you do not want to update before you initiate a cluster update.

Author: In this case since I'm talking about a feature/ability available to users, I think I'll replace "would" with "can".

> Because the MCO does not update nodes within paused MCPs, you would then pause the MCPs containing nodes that you do not want to update before you initiate a cluster update.

Reviewer 4: That's fine with me. It's not updated in the preview yet though, just fyi.

> Because the MCO does not update nodes within paused MCPs, you would then pause the MCPs containing nodes that you do not want to update before you initiate a cluster update.

Author: Good catch, I think I only applied the "initiating" part of your feedback and moved on without checking for the other part!
